### ISSUE TITLE
MFEM : Optimization flag being ignored

### ISSUE THREAD

**Issue by junkudo** [spack/spack#6745]
My command line optimization flag is being over-written by the MFEM build.
Spack command:
```
spack -d install -v mfem@3.3.2%intel@16.0.3 cflags="-g -O0" cppflags="-g -O0" cxxflags="-g -O0" fflags="-g -O0" ~debug~examples~gzstream+hypre+lapack~miniapps~mpfr+mpi+netcdf~openmp~petsc~suite-sparse~sundials~superlu-dist~threadsafe arch=linux-rhel7-x86_64
```

Spack debug output snippet:
```
[cc] /usr/tce/packages/intel/intel-16.0.3/bin/icpc -I/usr/tce/packages/mvapich2/mvapich2-2.3-intel-16.0.3/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/zlib-1.2.11-hhe7xrd6rlf7afevulaptctrum3673we/include -I/g/g\
14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/hypre-2.12.1-3asacyzbswxnimncl3jfad6gq2tizfap/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/metis-4.0.3-sacetyhy56rhqu6ggx3hzjfcwsaxs5t3/include -I/g/g14/\
kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/netcdf-4.4.1.1-6h45mookfhf7u5grqknfb5uc7dz2h5pz/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/hdf5-1.10.1-soxcwgh2bnzzqmq5bvhvplysjxtizrfx/include -g -O0 -g \
-O0 -g -O3 -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/hypre-2.12.1-3asacyzbswxnimncl3jfad6gq2tizfap/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/metis-4.0.3-sacetyhy56rhqu6ggx3hzjfcwsaxs5t3/\
include -I/usr/lib64/include -I/g/g14/kudo4/spack/opt/spack/linux-rhel7-x86_64/intel-16.0.3/netcdf-4.4.1.1-6h45mookfhf7u5grqknfb5uc7dz2h5pz/include -c general/array.cpp -o general/array.o -I/usr/tce/packages/mvapich2/mvapich2-2.3-intel-\
16.0.3/include
```
Note the `-O3` being introduced.

### Information on your system
Platform : TOSS3/TOSS (Tri-Lab Operating System Stack)

Relevant configuration details:
compilers.yaml
```
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tce/packages/intel/intel-16.0.3/bin/icc
      cxx: /usr/tce/packages/intel/intel-16.0.3/bin/icpc
      f77: /usr/tce/packages/intel/intel-16.0.3/bin/ifort
      fc: /usr/tce/packages/intel/intel-16.0.3/bin/ifort
    flags: {}
    spec: intel@16.0.3
    target: x86_64
```

packages.yaml:
```
packages:
  all:
    compiler: [intel, gcc, pgi, clang, xl, nag]
    providers:
      mpi: [mvapich2]
      blas: [netlib-lapack]
      lapack: [netlib-lapack]
  metis:
    version: [4.0.3]
  netlib-lapack:
    paths:
      netlib-lapack@3.4.2: /usr/lib64/
    buildable: False
  mvapich2:
    paths:
      mvapich2@2.3: /usr/tce/packages/mvapich2/mvapich2-2.3-intel-16.0.3/
    buildable: False
  mpich:
    buildable: False
  openmpi:
    buildable: False
  intel-mpi:
    buildable: False
  spectrum-mpi:
    buildable: False
  intel-parallel-studio:
    buildable: False
  cmake:
    paths:
      cmake@3.5.2 arch=linux-rhel7-x86_64: /usr/tce/packages/cmake/cmake-3.5.2/
    buildable: False
  pkg-config:
    paths:
      pkg-config@0.27.1 arch=linux-rhel7-x86_64: /usr/bin/
    buildable: False
  m4:
    paths:
      m4@1.4.16 arch=linux-rhe17-x86_64: /usr/bin/
    buildable: False
```

**Comment by adamjstewart** [comment]
See my comment [here](https://github.com/spack/spack/issues/6769#issuecomment-354323678). It's not that your flags are being ignored, it's that they are being handled by the compiler wrapper, not `CFLAGS` and friends.

**Comment by junkudo** [comment]
@adamjstewart could you re-open this issue based on the conversation in #6732?

What's the best way to fix this issue?  Since it's not a `cmake` configuration, I can't use the `flag_handler` approach used in PR #6714.

**Comment by v-dobrev** [comment]
Hi @junkudo,

Why not simply use the `+debug` variant of mfem? That will automatically switch the compiler options and, in addition, enable a lot of `MFEM_ASSERT` checks inside mfem. If you really want to be able to change the options, this can be done in mfem/package.py - just let me know and I can add it easily.

**Comment by junkudo** [comment]
Hey,
It's less about my specific optimization flags (`-O0` in this case) and more about how if anyone were to specify any optimization level flag that is not `-O3`, it would be over-written mfem's defaults.

**Comment by v-dobrev** [comment]
OK, I'll add support for the flags in the branch for PR #7423 - it already has a major update for the mfem package, so it will be a small addition there.

**Comment by v-dobrev** [comment]
This issue was addressed in #7423 which is now merged into `develop`.

I'll close the issue. Feel free to re-open if the issue is not fully resolved.
